engine/lib/worker/master_client.go

```diff
@@ -0,0 +1,247 @@
+	case err := <-errCh:
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *MasterClient) asyncReloadMasterInfo(ctx context.Context) <-chan error {
+	errCh := make(chan error, 1)
+	go func() {
+		defer close(errCh)
+		metaClient := metadata.NewMasterMetadataClient(m.masterID, m.frameMetaClient)
+		masterMeta, err := metaClient.Load(ctx)
+		if err != nil {
+			errCh <- err
+		}
+
+		m.putMasterInfo(masterMeta.NodeID, masterMeta.Epoch)
+	}()
+	return errCh
+}
+
+func (m *MasterClient) putMasterInfo(nodeID p2p.NodeID, epoch libModel.Epoch) {
+	m.infoLock.Lock()
+	defer m.infoLock.Unlock()
+
+	if epoch < m.masterEpoch {
+		return
+	}
+
+	if epoch >= m.masterEpoch {
+		m.masterEpoch = epoch
+		m.masterNode = nodeID
+	}
+}
+
+func (m *MasterClient) getMasterInfo() (p2p.NodeID, libModel.Epoch) {
+	m.infoLock.RLock()
+	defer m.infoLock.RUnlock()
+
+	return m.masterNode, m.masterEpoch
+}
+
+// MasterID returns the masterID.
+func (m *MasterClient) MasterID() libModel.MasterID {
+	// No need to lock here as masterID is immutable.
+	return m.masterID
+}
+
+// MasterNode returns the node ID of the executor
+// on which the master is being run.
+func (m *MasterClient) MasterNode() p2p.NodeID {
+	nodeID, _ := m.getMasterInfo()
+	return nodeID
+}
+
+// Epoch returns the master epoch.
+// Note that the epoch is increased when the master
+// restarts.
+func (m *MasterClient) Epoch() libModel.Epoch {
+	_, epoch := m.getMasterInfo()
+	return epoch
+}
+
+// HandleHeartbeat handles heartbeat messages received from the master.
+func (m *MasterClient) HandleHeartbeat(sender p2p.NodeID, msg *libModel.HeartbeatPongMessage) {
+	if msg.ToWorkerID != m.workerID {
+		log.L().Warn("Received heartbeat for wrong workerID",
+			zap.Any("msg", msg), zap.String("actual-worker-id", m.workerID))
+		return
+	}
+
+	_, epoch := m.getMasterInfo()
+	if msg.Epoch < epoch {
+		log.L().Info("epoch does not match, ignore stale heartbeat",
+			zap.Any("msg", msg),
+			zap.Int64("master-epoch", epoch))
+		return
+	}
+
+	if msg.Epoch > epoch {
+		// We received a heartbeat from a restarted master, we need to record
+		// its information.
+		m.putMasterInfo(sender, msg.Epoch)
+	}
+
+	if msg.IsFinished {
+		m.masterSideClosed.Store(true)
+	}
+	m.lastMasterAckedPingTime.Store(time.Duration(msg.SendTime))
+}
+
+// CheckMasterTimeout checks whether the master has timed out, i.e. we have lost
+// contact with the master for a while.
+func (m *MasterClient) CheckMasterTimeout(clk clock.Clock) (ok bool, err error) {
+	lastMasterAckedPingTime := clock.MonotonicTime(m.lastMasterAckedPingTime.Load())
+
+	sinceLastAcked := clk.Mono().Sub(lastMasterAckedPingTime)
+	if sinceLastAcked <= 2*m.timeoutConfig.WorkerHeartbeatInterval {
+		return true, nil
+	}
+
+	if sinceLastAcked > 2*m.timeoutConfig.WorkerHeartbeatInterval &&
+		sinceLastAcked < m.timeoutConfig.WorkerTimeoutDuration {
+
+		// We ignore the error here
+		_ = m.asyncReloadMasterInfo(context.Background())
```

Comment:
Is it better to add a timeout, such as `WorkerTimeoutDuration`, in case of leaving some stale goroutines even worker has exited.

Reply:
> Besides why removing the context parameter in this function

The semantics of `CheckMasterTimeout` does not seem to require a context. It should always be non-blocking to avoid any surprises.